### PR TITLE
[DA-778] Wangy70/implement biobank reconciliation reports

### PR DESCRIPTION
**Pull request by wangy70**
Remove 24hour report, change received and missing to cover 10 hours. Add a new modified report.

### CODE REVIEW

File: rest-api/offline/biobank_samples_pipeline.py [hunk 1]
```diff
@@ -207,43 +207,44 @@ def _get_report_paths(report_datetime):
...
-  # Open three files and a database session; run the reconciliation query and pipe the output
+  # Open two files and a database session; run the reconciliation query and pipe the output
   # to the files, using per-file predicates to filter out results.
-  with exporter.open_writer(path_received, received_predicate) as received_writer, \
-       exporter.open_writer(path_late, late_predicate) as late_writer, \
-       exporter.open_writer(path_missing, missing_predicate) as missing_writer, \
+  with exporter.open_writer(path_received, received_predicate) as received_writer,\
+       exporter.open_writer(path_missing, missing_predicate) as missing_writer,\
+       exporter.open_writer(path_modified, modified_predicate) as modified_writer,\
        database_factory.get_database().session() as session:
```
Comment (wangy70) [inline rest-api/offline/biobank_samples_pipeline.py]: add new modified report.

File: rest-api/offline/biobank_samples_pipeline.py [hunk 2]
```diff
@@ -253,11 +254,11 @@ def _query_and_write_reports(exporter, now, path_received, path_late, path_missi
...
                                       'tracking_number_system': _TRACKING_NUMBER_SYSTEM})
 
-  # Now generate the withdrawal report.
+  # Now generate the withdrawal report, within the past 10 days.
   exporter.run_export(path_withdrawals, replace_isodate(_WITHDRAWAL_REPORT_SQL),
                       {'race_question_code_id': race_question_code.codeId,
                        'native_american_race_code_id': native_american_race_code.codeId,
-                       'seven_days_ago': now - datetime.timedelta(days=7),
+                       'ten_days_ago': now - datetime.timedelta(days=10),
                        'biobank_id_prefix': get_biobank_id_prefix()})
```
Comment (wangy70) [inline rest-api/offline/biobank_samples_pipeline.py]: change withdrawals report to cover 10 days.

File: rest-api/offline/biobank_samples_pipeline.py [hunk 3]
```diff
@@ -297,6 +299,15 @@ def _query_and_write_reports(exporter, now, path_received, path_late, path_missi
...
+    LEFT OUTER JOIN
+      site restored_site
+    ON biobank_order.restored_site_id = restored_site.site_id
+    LEFT OUTER JOIN
+      site amended_site
+    ON biobank_order.amended_site_id = amended_site.site_id
+    LEFT OUTER JOIN
+      site cancelled_site
+    ON biobank_order.cancelled_site_id = cancelled_site.site_id
 """
```
Comment (wangy70) [inline rest-api/offline/biobank_samples_pipeline.py]: join these three for restored/amended/cancelled site names

File: rest-api/offline/biobank_samples_pipeline.py [hunk 4]
```diff
@@ -315,6 +326,48 @@ def _get_hpo_type_sql(hpo_alias):
...
+        WHEN biobank_order.order_status = {amended} OR biobank_order.order_status = {cancelled} OR
+             (biobank_order.order_status = {unset} AND biobank_order.restored_time IS NOT NULL)
+          THEN biobank_order.amended_reason
+        ELSE NULL  
+      END edited_cancelled_restored_site_reason
+  """.format(amended=int(BiobankOrderStatus.AMENDED), cancelled=int(BiobankOrderStatus.CANCELLED),
+             unset=int(BiobankOrderStatus.UNSET))
+
+  return result
+
```
Comment (wangy70) [inline rest-api/offline/biobank_samples_pipeline.py]: Add 5 new columns.

File: rest-api/offline/biobank_samples_pipeline.py [hunk 5]
```diff
@@ -479,5 +542,5 @@ def in_past_week(result, now, ordered_before=None):
     else:
       max_time = received_time
   if max_time:
-    return (now - max_time).days <= 7
+    return (now - max_time).days <= n_days
   return False
```
Comment (wangy70) [inline rest-api/offline/biobank_samples_pipeline.py]: change 'in_past_week' method to 'in_past_n_days'.

File: rest-api/test/unit_test/dao_test/biobank_order_dao_test.py [hunk 6]
```diff
@@ -319,3 +319,5 @@ def test_amending_an_order(self):
     self.assertEqual(amended_order.version, 2)
     self.assertEqual(amended_order.orderStatus, BiobankOrderStatus.AMENDED)
     self.assertEqual(amended_order.amendedReason, 'I had to change something')
+    self.assertEqual(amended_order.amendedSiteId, 1)
+    self.assertEqual(amended_order.amendedUsername, 'mike@pmi-ops.org')
```
Comment (wangy70) [inline rest-api/test/unit_test/dao_test/biobank_order_dao_test.py]: improve biobank_order_dao_test to make sure amended site and username are updated.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 7]
```diff
@@ -71,6 +76,74 @@ def _withdraw(self, participant, withdrawal_time):
       participant.withdrawalStatus = WithdrawalStatus.NO_USE
       self.participant_dao.update(participant)
 
+  def _modify_order(self, modify_type, order):
+    if modify_type == 'CANCELLED':
+      cancelled_request = self._get_cancel_patch()
+      self.order_dao.update_with_patch(order.biobankOrderId, cancelled_request, order.version)
```
Comment (wangy70) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: _modify_order method for test case

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 8]
```diff
@@ -160,9 +233,12 @@ def test_reconciliation_query(self):
...
-    self._insert_order(p_on_time, 'GoodOrder', BIOBANK_TESTS[:4], order_time,
-                       finalized_tests=BIOBANK_TESTS[:3], kit_id='kit1', tracking_number='t1',
-                       collected_note=u'\u2013foo', processed_note='bar', finalized_note='baz')
+    on_time_order = self._insert_order(p_on_time, 'GoodOrder', BIOBANK_TESTS[:4], order_time,
+                                       finalized_tests=BIOBANK_TESTS[:3], kit_id='kit1',
+                                       tracking_number='t1', collected_note=u'\u2013foo',
+                                       processed_note='bar', finalized_note='baz')
+    # edited order with matching sample; show both in rx and modified
+    self._modify_order('EDITED', on_time_order)
     self._insert_samples(p_on_time, BIOBANK_TESTS[:2], ['GoodSample1', 'GoodSample2'], 'OGoodOrder',
```
Comment (wangy70) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: test 'edited' order

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 9]
```diff
@@ -203,11 +279,12 @@ def test_reconciliation_query(self):
...
-    # On time order and samples from 10 days ago; shows up in rx
+    # On time order and samples from 10 days ago; should not shows up in rx or modified
     p_old_on_time = self._insert_participant(race_codes=[RACE_AIAN_CODE])
     # Old missing samples from 10 days ago don't show up in missing or late.
-    self._insert_order(p_old_on_time, 'OldGoodOrder', BIOBANK_TESTS[:3],
-                       old_order_time, kit_id='kit2')
+    old_on_time_order = self._insert_order(p_old_on_time, 'OldGoodOrder', BIOBANK_TESTS[:3],
+                                           old_order_time, kit_id='kit2')
+    self._modify_order('EDITED', old_on_time_order)
     self._insert_samples(p_old_on_time, BIOBANK_TESTS[:2], ['OldGoodSample1', 'OldGoodSample2'],
```
Comment (wangy70) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: test 'edited' order for more than 10 days, make sure they will not show in reports.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 10]
```diff
@@ -257,6 +333,22 @@ def test_reconciliation_query(self):
...
 
+    # cancelled/restored order with not matching sample; Does not show in rx, but should in modified
+    p_modified_on_time = self._insert_participant()
+    modified_order = self._insert_order(p_modified_on_time, 'CancelledOrder', BIOBANK_TESTS[:1],
+                                        order_time, finalized_tests=BIOBANK_TESTS[:1],
+                                        kit_id='kit6', tracking_number='t6',
+                                        collected_note=u'\u2013foo', processed_note='bar',
+                                        finalized_note='baz')
+    self._modify_order('CANCELLED', modified_order)
+
```
Comment (wangy70) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: test 'cancelled' order.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 11]
```diff
@@ -257,6 +333,22 @@ def test_reconciliation_query(self):
...
+    modified_order = self._insert_order(p_modified_on_time, 'RestoredOrder', BIOBANK_TESTS[:1],
+                                        order_time, finalized_tests=BIOBANK_TESTS[:1],
+                                        kit_id='kit7', tracking_number='t7',
+                                        collected_note=u'\u2013foo', processed_note='bar',
+                                        finalized_note='baz')
+    self._modify_order('RESTORED', modified_order)
+
```
Comment (wangy70) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: test 'restored' order.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 12]
```diff
@@ -464,6 +528,30 @@ def test_reconciliation_query(self):
         multi_sample_row['sent_order_id'].split(','),
         ['ORepeatedOrder2'])
 
+    # modified biobank orders show in modified report
+    exporter.assertRowCount(modified, 6)
+    exporter.assertHasRow(modified, {
+      'biobank_id': to_client_biobank_id(p_on_time.biobankId),
```
Comment (wangy70) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: verify modified report result.

**Comment by wangy70** [comment]
We don't have SQL injection risk for this one, please ignore the Codacy error: "Possible SQL injection vector through string-based query construction."

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 13]
```diff
@@ -71,6 +76,74 @@ def _withdraw(self, participant, withdrawal_time):
       participant.withdrawalStatus = WithdrawalStatus.NO_USE
       self.participant_dao.update(participant)
 
+  def _modify_order(self, modify_type, order):
+    if modify_type == 'CANCELLED':
+      cancelled_request = self._get_cancel_patch()
+      self.order_dao.update_with_patch(order.biobankOrderId, cancelled_request, order.version)
+    elif modify_type == 'EDITED':
```
Comment (Seeker1911) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: We should change edited to amended to remain consistent.

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 14]
```diff
@@ -203,11 +279,12 @@ def test_reconciliation_query(self):
                          'Ounconfirmed_sample_4', two_days_ago, two_days_ago)
 
 
-    # On time order and samples from 10 days ago; shows up in rx
+    # On time order and samples from 10 days ago; should not shows up in rx or modified
```
Comment (Seeker1911) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: shows -> show
At what time do received orders fall off of the received report ?

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 15]
```diff
@@ -203,11 +279,12 @@ def test_reconciliation_query(self):
...
     p_old_on_time = self._insert_participant(race_codes=[RACE_AIAN_CODE])
     # Old missing samples from 10 days ago don't show up in missing or late.
-    self._insert_order(p_old_on_time, 'OldGoodOrder', BIOBANK_TESTS[:3],
-                       old_order_time, kit_id='kit2')
+    old_on_time_order = self._insert_order(p_old_on_time, 'OldGoodOrder', BIOBANK_TESTS[:3],
+                                           old_order_time, kit_id='kit2')
+    self._modify_order('EDITED', old_on_time_order)
```
Comment (Seeker1911) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: edited -> amended

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py [hunk 16]
```diff
@@ -230,7 +306,7 @@ def test_reconciliation_query(self):
                         order_time,
                         order_time - datetime.timedelta(hours=1))
 
-    # Late order and samples from 10 days ago; shows up in rx (but not missing, as it was too
+    # Late order and samples from 10 days ago; should not shows up in rx and missing, as it was too
```
Comment (Seeker1911) [inline rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: shows -> show

File: rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py (hunk 14 above)
Comment (wangy70) [reply rest-api/test/unit_test/offline_test/biobank_samples_pipeline_mysql_test.py]: There is already a filter of 7days, but only applied to missing. I kept the logic and change it to 10days. Now it's applied to all reports.
The logic is `(now - max(received_time, collection_time)).days <= 10`. 

@Seeker1911 Other comments are all done, please review.

**Comment by wangy70** [comment]
@calbach Please give a final review, thanks!

File: rest-api/offline/biobank_samples_pipeline.py [hunk 17]
```diff
@@ -207,43 +207,44 @@ def _get_report_paths(report_datetime):
...
+  # Gets samples or orders where something has modified within the past 10 days.
+  modified_predicate = lambda result: (result[_EDITED_CANCELLED_RESTORED_STATUS_FLAG_INDEX] and
+                                       in_past_n_days(result, now, 10))
+
   code_dao = CodeDao()
   race_question_code = code_dao.get_code(PPI_SYSTEM, RACE_QUESTION_CODE)
   native_american_race_code = code_dao.get_code(PPI_SYSTEM, RACE_AIAN_CODE)
 
-  # Open three files and a database session; run the reconciliation query and pipe the output
+  # Open two files and a database session; run the reconciliation query and pipe the output
```
Comment (calbach) [inline rest-api/offline/biobank_samples_pipeline.py]: nit: still 3 right?

File: rest-api/offline/biobank_samples_pipeline.py [hunk 18]
```diff
@@ -315,6 +326,48 @@ def _get_hpo_type_sql(hpo_alias):
   result += "ELSE 'UNKNOWN' END)"
   return result
 
+
+def _get_status_flag_sql():
+  result = """
+      CASE
+        WHEN biobank_order.order_status = {amended} THEN 'edited'
```
Comment (calbach) [inline rest-api/offline/biobank_samples_pipeline.py]: optional: "edited" vs "amended" here (and throughout, e.g. column names)? Should it be "amended" for consistency?

Comment (wangy70) [reply rest-api/offline/biobank_samples_pipeline.py]: Actually the value "edited" and the column name is from the requirement document. That's why we have this unconsistency issue.

File: rest-api/offline/biobank_samples_pipeline.py [hunk 19]
```diff
@@ -207,43 +207,44 @@ def _get_report_paths(report_datetime):
...
+  modified_predicate = lambda result: (result[_EDITED_CANCELLED_RESTORED_STATUS_FLAG_INDEX] and
+                                       in_past_n_days(result, now, 10))
+
   code_dao = CodeDao()
   race_question_code = code_dao.get_code(PPI_SYSTEM, RACE_QUESTION_CODE)
   native_american_race_code = code_dao.get_code(PPI_SYSTEM, RACE_AIAN_CODE)
 
   # Open three files and a database session; run the reconciliation query and pipe the output
   # to the files, using per-file predicates to filter out results.
-  with exporter.open_writer(path_received, received_predicate) as received_writer, \
```
Comment (wangy70) [inline rest-api/offline/biobank_samples_pipeline.py]: Yes, it should be three.

**Pull request merged**